pulsar-grpc/src/main/java/org/apache/pulsar/grpc/ConsumerHandler.java

```diff
@@ -0,0 +1,241 @@
+package org.apache.pulsar.grpc;
+
+import com.google.common.base.Enums;
+import com.google.common.base.Strings;
+import com.google.protobuf.ByteString;
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import org.apache.pulsar.client.api.*;
+import org.apache.pulsar.client.impl.ConsumerBuilderImpl;
+import org.apache.pulsar.common.naming.TopicName;
+import org.apache.pulsar.grpc.proto.ConsumerAck;
+import org.apache.pulsar.grpc.proto.ConsumerMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static org.apache.pulsar.grpc.Constant.GRPC_PROXY_CTX_KEY;
+
+public class ConsumerHandler implements Closeable {
+
+    private static final Logger log = LoggerFactory.getLogger(ConsumerHandler.class);
+
+    private final GrpcService service;
+    private final StreamObserver<ConsumerMessage> messageStreamObserver;
+    private String topic;
+    private String subscription;
+    private Consumer<byte[]> consumer;
+
+    private int maxPendingMessages;
+    private final AtomicInteger pendingMessages = new AtomicInteger();
+
+    private volatile long msgDeliveredCounter = 0;
+    private static final AtomicLongFieldUpdater<ConsumerHandler> MSG_DELIVERED_COUNTER_UPDATER =
+            AtomicLongFieldUpdater.newUpdater(ConsumerHandler.class, "msgDeliveredCounter");
+
+    ConsumerHandler(GrpcService service, StreamObserver<ConsumerMessage> messageStreamObserver) {
+        this.service = service;
+        this.messageStreamObserver = messageStreamObserver;
+        Map<String, String> params = GRPC_PROXY_CTX_KEY.get();
```

Comment:
can you explain how does this work? it seems to me that `GRPC_PROXY_CTX_KEY` is a static field and shared across different streams. how does each stream know its own contxt?

Reply:
The key is static but the `get()` on the key will get the value for the request scoped Context.